### PR TITLE
Codecs and combinators

### PR DESCRIPTION
**Pull request by gusty**

### CODE REVIEW

**Comment by gusty** [comment]
I am not sure this is the best way to add combinators.
Using this kind of static member extensions is limited to non-generics, otherwise it becomes quickly unusable. For generics we use the typical module of functions.

The drawback of this is that is not easier discoverable. For instance to deserialize a `Choose<_,_,_>` you need to know you need to type `Choose3.ofJsonWith`.

The other problem is when using 2 implementations at the same time, names will clash.

The alternative is the classical `Decode` module, which is used in most Elm inspired libraries.

Here, we could use the same class (or better a module with the same name) we use already to store the overloads, for example:

`OfJson.choose3`

**Comment by wallymathieu** [comment]
Having an `OfJson` module for each implementation sounds like a good idea.

File: Fleece/Fleece.fs [hunk 1]
```diff
@@ -664,9 +677,7 @@ module SystemJson =
     type OfJson with 
         static member inline OfJson (_: 'R, _:Default4) =
             let codec = (^R : (static member JsonObjCodec: Codec<IReadOnlyDictionary<string,JsonValue>,'R>) ())
-            function
-            | JObject o -> decode (fst codec) o : ^R ParseResult
-            | a         -> failparse "Map" a
+            codec |> Codec.compose jsonObjToValueCodec |> fst : JsonValue -> ^R ParseResult
```
Comment (wallymathieu) [inline Fleece/Fleece.fs]: looks good, tells more about what it does

**Comment by gusty** [comment]
This closes #4

**Pull request merged**